source/non_matching/mesh_classifier.cc

```diff
@@ -0,0 +1,541 @@
+    : triangulation(&dof_handler.get_triangulation())
+    , level_set_description(
+        std::make_unique<DiscreteLevelSetDescription<dim, VECTOR>>(dof_handler,
+                                                                   level_set))
+  {}
+
+
+
+  template <int dim>
+  MeshClassifier<dim>::MeshClassifier(const Triangulation<dim> &triangulation,
+                                      const Function<dim> &     level_set,
+                                      const FiniteElement<dim> &element)
+    : triangulation(&triangulation)
+    , level_set_description(
+        std::make_unique<AnalyticLevelSetDescription<dim>>(level_set, element))
+  {
+    // The level set function must be scalar.
+    AssertDimension(level_set.n_components, 1);
+    AssertDimension(element.n_components(), 1);
+  }
+
+
+
+  template <int dim>
+  void
+  MeshClassifier<dim>::reclassify()
+  {
+    initialize();
+
+    // Loop over all cells and determine category of all locally owned
+    // cells and faces.
+    for (const auto &cell : triangulation->active_cell_iterators())
+      {
+        if (cell->is_locally_owned())
+          {
+            for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell; ++f)
+              {
+                face_locations[cell->face(f)] =
+                  determine_face_location_to_levelset(cell, f);
+              }
+
+            cell_locations[cell] = determine_cell_location_to_levelset(cell);
+          }
+      }
+  }
+
+
+
+  template <int dim>
+  LocationToLevelSet
+  MeshClassifier<dim>::determine_cell_location_to_levelset(
+    const typename Triangulation<dim>::active_cell_iterator &cell)
+  {
+    const unsigned int fe_index = level_set_description->active_fe_index(cell);
+
+    const unsigned int n_local_dofs = lagrange_to_bernstein_cell[fe_index].m();
+
+    Vector<double> local_levelset_values(n_local_dofs);
+    level_set_description->get_local_level_set_values(cell,
+                                                      local_levelset_values);
+
+    lagrange_to_bernstein_cell[fe_index].solve(local_levelset_values);
+
+    return location_from_dof_signs(local_levelset_values);
+  }
+
+
+
+  template <int dim>
+  LocationToLevelSet
+  MeshClassifier<dim>::determine_face_location_to_levelset(
+    const typename Triangulation<dim>::active_cell_iterator &cell,
+    const unsigned int                                       face_index)
+  {
+    const unsigned int fe_index = level_set_description->active_fe_index(cell);
+
+    const unsigned int n_local_dofs =
+      lagrange_to_bernstein_face[fe_index][face_index].m();
+
+    Vector<double> local_levelset_values(n_local_dofs);
+    level_set_description->get_local_level_set_values(cell,
+                                                      face_index,
+                                                      local_levelset_values);
+
+    lagrange_to_bernstein_face[fe_index][face_index].solve(
+      local_levelset_values);
```

Comment:
As a future improvement not necessary for this PR we might consider to switch this to tensor product computations, again with the aim to reduce the computation cost. If I read the code correctly, this transformation to the Bernstein basis preserves the tensor product of the 1D shape functions in `FE_Q` and `FE_Bernstein`, so all we need is the 1D version and some of the function `do_forward` in https://github.com/dealii/dealii/blob/032a42995a4dd6a2d4e1b212ae50bc540527b764/include/deal.II/matrix_free/evaluation_kernels.h#L843-L850